### ISSUE TITLE
Local cross-encoder rerank is ~5.7 s per 30-document window and does not scale with threads (8 → 32 threads: no change)

### ISSUE THREAD

**Issue by xerj-org** [xerj-org/xerj#964]
## Measured

Branch `feat/local-judge`, `pair_score` example (same model runtime, tokenizer settings, length-aware batching and thread pool as the in-server `local` rerank provider — see its module doc), one real 30-pair window from `scifact.test.w30.jsonl`, box idle apart from this (load ~6 of 32 cores):

| tier | model | 8 threads | 32 threads | 30-doc window |
|---|---|---:|---:|---:|
| `small` | cross-encoder/ms-marco-MiniLM-L6-v2 (90 MB) | 5.28 pairs/s | 5.20 pairs/s | **~5.7 s** |
| `base` | BAAI/bge-reranker-base | 1.91 pairs/s | 1.97 pairs/s | **~15.4 s** |

Four times the threads changes throughput by less than 2%, in both directions — so the thread pool is not the limit, and nothing is scaling.

## Why it matters

A rerank stage runs inside a search request. At ~5.7 s for the smallest tier the local provider cannot serve interactive search, and `base` is three times worse. For scale: the hosted judge this provider is an alternative to publishes a p95 of 0.8–1.8 s for a 30-document call (hev/jev-rerank's README — published by them, not run by us).

The quality is there — on NFCorpus, hybrid top-30 → `local small` measures nDCG@10 0.3597 against 0.3445 for hybrid alone (+0.0152, 95% CI [+0.0047, +0.0268]) — so latency, not relevance, is what stands between this and a shippable local judge.

## Likely the same root cause as #938

#938 measured neural *ingest* keeping ~3.4 of 32 threads busy: 6.1 docs/s from one `_bulk` stream against 29.5 docs/s from 8 concurrent clients for the same CPU-seconds. Same shape here: one caller cannot saturate the machine, and more threads do not help. Worth investigating together — if a single forward pass is effectively serial, both the embedder and the cross-encoder inherit it.

## Reproduce

```sh
cd engine && cargo build --release -p xerj-ai --features neural --example pair_score
# one window = 30 lines sharing the same "g" from a *.w30.jsonl pair file
pair_score --tier small --threads 8  --in win.jsonl --out /tmp/o.jsonl --summary /tmp/s.json
pair_score --tier small --threads 32 --in win.jsonl --out /tmp/o2.jsonl --summary /tmp/s2.json
```

`pairs_per_second` in the summary is the number above. Harness and first-stage data: `benchmarks/local-judge/` on that branch.

**Comment by xerj-org** [comment]
Fixed on `feat/local-judge` @ e7a8bfd3 (pushed).

**Root cause.** One window is one serial chain of forward passes. `CHUNK_ROWS=32` makes a 30-pair window a single chunk, `group_by_padded_cost` emits 1–2 batches under the 4,096-slot budget, and `logits_blocking` runs them one at a time — the judge pool wrapped around it by `pool.install` has nothing for its other threads to do. Candle cannot absorb the width either: gemm's thread count is process-global, and the per-pass chain of gemms + single-threaded elementwise ops caps one pass at a few cores. Two doc comments claimed the opposite; they are corrected.

**Fix.** `logits_blocking` now takes the split width from `rayon::current_num_threads()` (so `[judge] threads` already controls it), re-cuts the chunk's plan with a `PADDED_TOKEN_BUDGET/width` per-pass budget, and runs up to width passes concurrently per wave. New `microbatch::plan_waves` packs consecutive batches into waves bounded two ways — at most width passes and at most 4,096 padded slots in flight — so the activation-memory ceiling is the same at every width. `width=1` reproduces the old serial pass sequence bit-identically.

**Measured** (same harness, pinned ms-marco-MiniLM-L6-v2, same 50 scifact windows, before/after the same afternoon; raw log in `benchmarks/local-judge/results/latency-964.txt`):

| threads | p50 before | p50 after | pairs/s before → after |
|---|---|---|---|
| 1 | 7,647 ms | 7,297 ms | 3.95 → 4.15 |
| 8 | 5,031 ms | 1,027 ms | 6.01 → 27.53 |
| 16 | — | 788 ms | — → 35.16 |
| 32 | 4,916 ms | 861 ms | 6.07 → 32.46 |

8 threads: **4.9× per window**; the auto width (16) lands at 788 ms p50 / 1,455 ms p95; the knee is at 16 (32 is slower again, peak RSS 818 → 1,105 MB). Logits over all 1,500 pairs vs the pre-fix binary: threads=1 bit-identical, threads=8/32 max |Δ| 5.7e-06 (175× under the existing 1e-3 tolerance), ranking order unchanged. FAIL-BEFORE proven; `xerj-ai` + `xerj-rerank` (local feature) green; clippy/fmt clean; ES-YAML 1371/0/3 on a private port.

**Status, honestly.** The local cross-encoder provider this issue measured is the opt-in provider on `feat/local-judge`, which rc.75 deliberately left off `main` (it does not beat the shipped hybrid RRF on SciFact; NFCorpus-only gain — see ROADMAP *zero-token direction*). This fix rides that branch and lands with it when it merges; it does not change the parked-not-default decision. The branch is 8 main-commits behind (rc.75 landed after its last merge point) — updating it is follow-up work, tracked here as done-where-the-code-lives.
